### PR TITLE
chore: Improved error messages for sign-in flow

### LINKED ISSUE #1091: Improvements to error messaging
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!

___

View the job in Upwork here. 
**Platform - version:**  ALL 

----
**Action Performed (reproducible steps):**
Summary: Currently, when a user encounters a sign-in error on expensify.cash, we display the raw API message to the user. We'd like to replace these responses so that when an error occurs, we display a user-friendly message instead. 

The error codes and associated user-friendly messages are included in the `Messaging` section of this issue. 

**Steps to reproduce an example of the issue:**
1. Go to expensify.cash
2. Sign in using an incorrect email address and password combination (i.e. a pair of credentials not associated with an account)
3. See error code and generic error message appear (in red text). 
     - Example: `401 No user with that partner/user/password`

**Desired Result:**
Show user-friendly error messages. Example: 401 error should show: `Incorrect password`

**Actual Result**
Generic  `401 No user with that partner/user/password` appears 

**Notes/Photos/Videos:** 
image
 
----

**Messaging:**

| Error Code  | Message to display |
| ------------- | ------------- |
| 401  | Incorrect login or password. Please try again. |
| 402  | You have 2FA enabled on this account. Please sign in using your email or phone number. |
| 403 | Invalid login or password. Please try again or reset your password. | 
| 404 | We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes. | 
| 405 |  You do not have access to this application. Please add your GitHub username for access. |

### PR DESCRIPTION
**Pull request by twointo2**
### Details
Currently, when a user encounters a sign-in error on [expensify.cash](https://expensify.cash/), we display the raw API message to the user. In this PR, I have made a utility function to extract the `jsonCode` from the API error response and display the respective error message based on the `jsonCode` defined in the newly added utility function `getErrorMessageFromErrorCode` defined in a new file `src/libs/ErrorUtils.js`.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #1091

### Tests
1. Open the Expensify application.
2. `Signup for Waitlist` screen will appear
3. Enter your Email/Phone into the text field
4. Tap on the `Continue Button`
5. `Enter Password` screen will appear
6. Try entering an invalid password
7. A red text with the message `Incorrect login or password. Please try again.` will appear
8. Repeat the same step for all of different use cases defined in the issue #1091 statement

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
Web

#### Mobile Web
Screenshot 2021-01-10 at 9 29 29 AM

#### Desktop
Desktop

#### iOS
Simulator Screen Shot - iPhone 11 Pro - 2021-01-09 at 11 04 03

#### Android
Screenshot_1610170603

### CODE REVIEW

**Comment by ajay426** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by puneetlath** [comment]
Can we make it “too many” instead of “so many”?

Puneet Lath

**Comment by ajay426** [comment]
Sure @puneetlath 
Now I have made it like:
`Your account has been locked after too many unsuccessfull attempts. Please try again after 1 hour.`

File: src/libs/ErrorUtils.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            return 'You have 2FA enabled on this account. Please sign in using your email or phone number.';
+        case 403:
+            return 'Invalid login or password. Please try again or reset your password.';
+        case 404:
+            // eslint-disable-next-line max-len
+            return 'We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes.';
+        case 405:
+            return 'You do not have access to this application. Please add your GitHub username for access.';
+        case 413:
+            return 'Your account has been locked after too many unsuccessfull attempts. Please try again after 1 hour.';
```
Comment (nickmurray47) [inline src/libs/ErrorUtils.js:26]: `unsuccessfull` -> `unsuccessful`

**Review by nickmurray47 — CHANGES_REQUESTED** [review]

**Comment by nickmurray47** [comment]
@ajay426 your solution looks good! There's one small spelling change to make, and could you also post screenshots of the other error messages (I only see two examples here)?

File: src/libs/ErrorUtils.js (hunk 1 above)
Comment (ajay426) [reply src/libs/ErrorUtils.js:26]: Got it 👍

**Comment by ajay426** [comment]
Sure @nickmurray47

**Comment by ajay426** [comment]
I would need the steps to produce these use-case so that I can reproduce the other use cases and capture a screenshot for them. I tried to find the use-case by myself, but was not able find the steps. I have also commented on the respective issue #1091 as well.

File: src/libs/ErrorUtils.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * e.g.
+ *
+ * getErrorMessageFromErrorCode(401)  =>  Incorrect login or password. Please try again.
+ *
+ * @param {Number} errorCode
+ *
+ * @returns {String}
+ */
+function getErrorMessageFromErrorCode(errorCode) {
```
Comment (yuwenmemon) [inline src/libs/ErrorUtils.js:12]: I would just place this in the context of `Authenticate`. Placing this here implies these error messages and codes are generic. They're not. They're only relevant in the context of `Authenticate`

**Review by yuwenmemon — CHANGES_REQUESTED** [review]

**Comment by nickmurray47** [comment]
> I would need the steps to produce these use-case so that I can reproduce the other use cases and capture a screenshot for them. I tried to find the use-case by myself, but was not able find the steps. I have also commented on the respective issue #1091 as well.

@ajay426 I was just going to suggest that you manually set the `response code` in the `Authenticate` function to simulate the other use cases, i.e. `const errorMessage = 405;`

**Comment by ajay426** [comment]
@yuwenmemon  @nickmurray47 
Do you mean that I should shift all of the `getErrorMessageFromErrorCode` logic into the `Authenticate` method in the `src/libs/API.js` file?
If that is the case, then I have 2 options:

### Option 1: Shift all logic into the `then` block of the `request` method like:
```
if (response.jsonCode !== 200) {
    let errorMessage = 'Something went wrong. Please try again later.';
    if (response.jsonCode === 401) {
        errorMessage = 'Incorrect login or password. Please try again.';
    }
    if (response.jsonCode === 402) {
        errorMessage = 'You have 2FA enabled on this account. Please sign in using your email or phone number.';
    }
    if (response.jsonCode === 403) {
        errorMessage = 'Invalid login or password. Please try again or reset your password.';
    }
    if (response.jsonCode === 404) {
        errorMessage = 'We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes.';
    }
    if (response.jsonCode === 405) {
        errorMessage = 'You do not have access to this application. Please add your GitHub username for access.';
    }
    if (response.jsonCode === 413) {
        errorMessage = 'Your account has been locked after too many unsuccessful attempts. Please try again after 1 hour.';
    }

    throw new Error(errorMessage);
}
```

### Option 2: 
Or may be, I can shift the whole function into the `API.js` file to keep the error message extraction logic separate and reduce the mess in the code. I can rename the function name in this case like: `getAuthErrorMessageFromErrorCode`.
What are your views guys?

**Comment by nickmurray47** [comment]
> Do you mean that I should shift all of the `getErrorMessageFromErrorCode` logic into the `Authenticate` method in the `src/libs/API.js` file?

@ajay426 I'd much prefer option 1 because it's the simplest, and there's not really a need for another function since these error codes will only be used within `Authenticate`

**Comment by yuwenmemon** [comment]
Options 1 looks fine, but use a `switch` statement please.

**Comment by ajay426** [comment]
Hey guys!
I have pushed the requested changes. Please check now.

File: src/libs/API.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable max-len */
```
Comment (nickmurray47) [inline src/libs/API.js:1]: Can you place `disable` line just for the `next-line` where it's needed rather than disable it for the whole file?

Comment (ajay426) [reply src/libs/API.js:1]: Sure.

Comment (ajay426) [reply src/libs/API.js]: Done. Please check now.

**Comment by ajay426** [comment]
Hi guys!
Here are the screenshots for the other use cases(I have simulated the error codes in the JS)

Screenshot 2021-01-16 at 9 56 15 AM
Screenshot 2021-01-16 at 9 56 36 AM
Screenshot 2021-01-16 at 9 57 03 AM
Screenshot 2021-01-16 at 9 57 20 AM
Screenshot 2021-01-16 at 10 01 21 AM

**Review by nickmurray47 — COMMENTED** [review]
@ajay426 do you mind merging with master as well? The code looks good to me! Thanks for making those changes. Since there's another contributor with a PR for the same issue, I just want to confirm with our team that we're going to move forward with your solution before approving and merging.

**Comment by ajay426** [comment]
Sure @nickmurray47. 

> do you mind merging with master as well?

Sorry! Didn't get this point. Are you asking me to merge the changes into the `master`? If so, then how can I do that? Can you please guide me in this?

**Review by yuwenmemon — APPROVED** [review]

**Review by nickmurray47 — APPROVED** [review]

**Pull request merged**